src/infrasys/chronify_time_series_storage.py

```diff
@@ -0,0 +1,346 @@
+"""Implementation of arrow storage for time series."""
+
+import atexit
+from contextlib import contextmanager
+from datetime import datetime, timedelta
+from functools import singledispatch
+from pathlib import Path
+from tempfile import NamedTemporaryFile
+from typing import Any, Generator, Self
+from uuid import UUID
+
+import pandas as pd
+import pint
+from chronify import DatetimeRange, Store, TableSchema
+from loguru import logger
+from sqlalchemy import Connection
+
+from infrasys.id_manager import IDManager
+from infrasys.time_series_models import (
+    SingleTimeSeries,
+    SingleTimeSeriesKey,
+    SingleTimeSeriesMetadata,
+    TimeSeriesData,
+    TimeSeriesKey,
+    TimeSeriesMetadata,
+    TimeSeriesStorageType,
+)
+from infrasys.time_series_storage_base import TimeSeriesStorageBase
+from infrasys.utils.path_utils import delete_if_exists
+
+
+_SINGLE_TIME_SERIES_BASE_NAME = "single_time_series"
+_TIME_SERIES_FILENAME = "time_series_data.db"
+
+
+class ChronifyTimeSeriesStorage(TimeSeriesStorageBase):
+    """Stores time series in a chronfiy database."""
+
+    def __init__(
+        self,
+        store: Store,
+        id_manager: IDManager,
+        read_only: bool = False,
+        uuid_lookup: dict[UUID, int] | None = None,
+    ) -> None:
+        self._store = store
+        self._read_only = read_only
+        # infrasys currently uses UUIDs as unique identifies for components and time series.
+        # Those will eventually use integer IDs instead.
+        # We don't want to store UUIDs in the chronify database.
+        # Integer IDs are much smaller and faster for search.
+        # Manage a mapping of UUIDs to integer IDs until we can remove UUIDs.
+        self._uuid_lookup: dict[UUID, int] = uuid_lookup or {}
+        self._id_manager = id_manager
+
+    @classmethod
+    def create_with_temp_directory(
+        cls,
+        base_directory: Path | None = None,
+        engine_name: str = "duckdb",
+        read_only: bool = False,
+    ) -> Self:
+        """Construct ChronifyTimeSeriesStorage with a temporary directory."""
+        with NamedTemporaryFile(dir=base_directory, suffix=".db") as f:
+            dst_file = Path(f.name)
+        logger.debug("Creating database at {}", dst_file)
+        atexit.register(delete_if_exists, dst_file)
+        store = Store(engine_name=engine_name, file_path=dst_file)
+        id_manager = IDManager(next_id=1)
+        return cls(store, id_manager, read_only=read_only)
+
+    @classmethod
+    def create_with_permanent_directory(
+        cls,
+        base_directory: Path,
+        engine_name: str = "duckdb",
+        read_only: bool = False,
+    ) -> Self:
+        """Construct ChronifyTimeSeriesStorage with a permanent directory."""
+        dst_file = base_directory / _TIME_SERIES_FILENAME
+        if dst_file.exists():
+            msg = f"Bug: time series database already exists: {dst_file}"
+            raise Exception(msg)
```

Comment:
Do we want to raise one of our custom exceptions instead?

Reply:
Think of these cases like asserts. Asserts can be optimized out, so they aren't guaranteed. If we fall into these conditionals, the code has a bug and a hard crash is needed.